### PR TITLE
Ditto integration

### PR DESCRIPTION
**Pull request by nicnys-8**
We would like this pull request to be merged into a new branch named ditto-integration.

### CODE REVIEW

File: core-common/src/main/java/eu/arrowhead/common/dto/internal/InventoryIdDTO.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+	// =================================================================================================
+	// members
+
+	private String inventoryId;
+
+	// =================================================================================================
+	// methods
+
+	// -------------------------------------------------------------------------------------------------
+	public InventoryIdDTO(@JsonProperty("inventoryId") final String inventoryId) {
```
Comment (rbocsi) [inline core-common/src/main/java/eu/arrowhead/common/dto/internal/InventoryIdDTO.java:28]: maybe a  `@JsonCreator` annotation on the constructor is missing here?

File: core-common/src/main/java/eu/arrowhead/common/dto/internal/SystemDataDTO.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+	// =================================================================================================
+	// members
+
+	private String data;
+
+	// =================================================================================================
+	// methods
+
+	// -------------------------------------------------------------------------------------------------
+	public SystemDataDTO(@JsonProperty("systemData") final String data) {
```
Comment (rbocsi) [inline core-common/src/main/java/eu/arrowhead/common/dto/internal/SystemDataDTO.java:28]: maybe a @JsonCreator annotation on the constructor is missing here?

File: core-common/src/main/java/eu/arrowhead/common/dto/internal/SystemDataDTO.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+		this.data = data;
+	}
+
+	// -------------------------------------------------------------------------------------------------
+	public String getSystemData() {
```
Comment (rbocsi) [inline core-common/src/main/java/eu/arrowhead/common/dto/internal/SystemDataDTO.java:33]: getSystemData and setSystemData() or getData() and setData().

File: ditto/src/deb/control/config [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+  USER=$(logname)
+  HOME=$(eval echo "~$USER")
+else
+  USER="root"
+  HOME="/root"
+fi
+
+# Get installation type
+db_input high arrowhead-core-common/db_host || true
+db_input high arrowhead-core-common/mysql_password_chor || true
```
Comment (rbocsi) [inline ditto/src/deb/control/config:15]: This field is the db password of the Choreographer. If you want to use the database, creates a new entry in the core-common/src/deb/control/templates file and use that

File: ditto/src/deb/control/postinst [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+    systems="$(ah_get_conf_prop systems)"
+    if [[ "${systems}" != *"${SYSTEM_NAME}"* ]]; then
+      systems="${systems}${SYSTEM_NAME} "
+    fi
+    if [[ "${systems}" != *"sysop"* ]]; then
+      systems="${systems}sysop "
+    fi
+    ah_set_conf_prop systems "${systems}"
+
+    db_get arrowhead-core-common/mysql_password_ditto; system_passwd=$RET || true
```
Comment (rbocsi) [inline ditto/src/deb/control/postinst:42]: in here, you use the right field (but this field is not defined in the templates file)

File: ditto/src/main/java/eu/arrowhead/core/ditto/DittoController.java [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+			@ApiResponse(code = HttpStatus.SC_INTERNAL_SERVER_ERROR, message = CoreCommonConstants.SWAGGER_HTTP_500_MESSAGE)
+	})
+	@GetMapping(path = CommonConstants.ECHO_URI)
+	public String echoService() {
+		return "Got it!";
+	}
+
+	//-------------------------------------------------------------------------------------------------
+	@ApiOperation(value = "Forward the given request to the appropriate Ditto Thing",
+			response = String.class, tags = {CoreCommonConstants.SWAGGER_TAG_CLIENT})
```
Comment (rbocsi) [inline ditto/src/main/java/eu/arrowhead/core/ditto/DittoController.java:62]: Is response a String or and Object?

File: ditto/src/main/java/eu/arrowhead/core/ditto/DittoMonitorController.java [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+	//-------------------------------------------------------------------------------------------------
+	@ApiOperation(value = "Return system data describing this core system", response = String.class, tags = { CoreCommonConstants.SWAGGER_TAG_CLIENT })
+	@ApiResponses(value = {
+			@ApiResponse(code = HttpStatus.SC_OK, message = CoreCommonConstants.SWAGGER_HTTP_200_MESSAGE),
+			@ApiResponse(code = HttpStatus.SC_UNAUTHORIZED, message = CoreCommonConstants.SWAGGER_HTTP_401_MESSAGE),
+			@ApiResponse(code = HttpStatus.SC_INTERNAL_SERVER_ERROR, message = CoreCommonConstants.SWAGGER_HTTP_500_MESSAGE)
+	})
+	@GetMapping(path = CommonConstants.SYSTEM_DATA_URI)
+	public SystemDataDTO getSystemData() {
+		return new SystemDataDTO("");
```
Comment (rbocsi) [inline ditto/src/main/java/eu/arrowhead/core/ditto/DittoMonitorController.java:54]: does not seem to me very useful.

File: ditto/src/main/java/eu/arrowhead/core/ditto/DittoMonitorController.java [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+			@ApiResponse(code = HttpStatus.SC_UNAUTHORIZED, message = CoreCommonConstants.SWAGGER_HTTP_401_MESSAGE),
+			@ApiResponse(code = HttpStatus.SC_INTERNAL_SERVER_ERROR, message = CoreCommonConstants.SWAGGER_HTTP_500_MESSAGE)
+	})
+	@GetMapping(path = CommonConstants.PING_URI)
+	public String monitorPing() {
+		return "OK";
+	}
+
+	//-------------------------------------------------------------------------------------------------
+	@ApiOperation(value = "Return system data describing this core system", response = String.class, tags = { CoreCommonConstants.SWAGGER_TAG_CLIENT })
```
Comment (rbocsi) [inline ditto/src/main/java/eu/arrowhead/core/ditto/DittoMonitorController.java:46]: response is not String

File: ditto/src/main/java/eu/arrowhead/core/ditto/DittoMonitorController.java [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+			@ApiResponse(code = HttpStatus.SC_UNAUTHORIZED, message = CoreCommonConstants.SWAGGER_HTTP_401_MESSAGE),
+			@ApiResponse(code = HttpStatus.SC_INTERNAL_SERVER_ERROR, message = CoreCommonConstants.SWAGGER_HTTP_500_MESSAGE)
+	})
+	@GetMapping(path = CommonConstants.SYSTEM_DATA_URI)
+	public SystemDataDTO getSystemData() {
+		return new SystemDataDTO("");
+	}
+
+	//-------------------------------------------------------------------------------------------------
+	@ApiOperation(value = "Return an inventory ID for this core system", response = String.class, tags = { CoreCommonConstants.SWAGGER_TAG_CLIENT })
```
Comment (rbocsi) [inline ditto/src/main/java/eu/arrowhead/core/ditto/DittoMonitorController.java:58]: response is not String

File: ditto/src/main/java/eu/arrowhead/core/ditto/DittoMonitorController.java [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+	@ApiResponses(value = {
+			@ApiResponse(code = HttpStatus.SC_OK, message = CoreCommonConstants.SWAGGER_HTTP_200_MESSAGE),
+			@ApiResponse(code = HttpStatus.SC_UNAUTHORIZED, message = CoreCommonConstants.SWAGGER_HTTP_401_MESSAGE),
+			@ApiResponse(code = HttpStatus.SC_INTERNAL_SERVER_ERROR, message = CoreCommonConstants.SWAGGER_HTTP_500_MESSAGE)
+	})
+	@GetMapping(path = CommonConstants.INVENTORY_ID_URI)
+	public InventoryIdDTO getInventoryId() {
+		return new InventoryIdDTO("");
```
Comment (rbocsi) [inline ditto/src/main/java/eu/arrowhead/core/ditto/DittoMonitorController.java:66]: does not seem to me very useful.

File: ditto/src/main/java/eu/arrowhead/core/ditto/security/DittoSecurityConfig.java [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+	//-------------------------------------------------------------------------------------------------
+	@Bean
+	public FilterRegistrationBean<DittoSecurityFilter> filterRegistrationBean() {
+		FilterRegistrationBean<DittoSecurityFilter> registrationBean =
+				new FilterRegistrationBean<DittoSecurityFilter>();
+		DittoSecurityFilter dittoSecurityFilter = new DittoSecurityFilter(arrowheadContext, httpService);
+
+		dittoSecurityFilter.setMyPrivateKey(getMyPrivateKey());
+		registrationBean.setFilter(dittoSecurityFilter);
+		registrationBean.addUrlPatterns("/things/*"); // TODO: Move constant somewhere else.
```
Comment (rbocsi) [inline ditto/src/main/java/eu/arrowhead/core/ditto/security/DittoSecurityConfig.java:62]: can be a constant in this class if you don't use anywhere else

File: ditto/src/main/java/eu/arrowhead/core/ditto/security/DittoSecurityConfig.java [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+		return registrationBean;
+	}
+
+	// -------------------------------------------------------------------------------------------------
+	private PrivateKey getMyPrivateKey() {
+		KeyStore keystore;
```
Comment (rbocsi) [inline ditto/src/main/java/eu/arrowhead/core/ditto/security/DittoSecurityConfig.java:68]: your private key can be accessed through the arrowheadContext too with the key CommonConstants.SERVER_PRIVATE_KEY.

See ApplicationInitListener.obtainKeys() method

File: ditto/src/main/resources/application.properties [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+# Show all request/response in debug log
+log_all_request_and_response=false
+
+# Service Registry web-server parameters (to register the Authorization services)
+sr_address=127.0.0.1
+sr_port=8443
+
+# Ditto parameters
+ditto_http_address=http://localhost:8080
+ditto_ws_address=ws://localhost:8080
```
Comment (rbocsi) [inline ditto/src/main/resources/application.properties:40]: two protocols on the same port?

File: ditto/src/test/resources/application.properties [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+# Show all request/response in debug log
+log_all_request_and_response=false
+
+# Service Registry web-server parameters (to register the Authorization services)
+sr_address=127.0.0.1
+sr_port=8443
+
+# Ditto parameters
+ditto_http_address=http://localhost:8080
+ditto_ws_address=ws://localhost:8080
```
Comment (rbocsi) [inline ditto/src/test/resources/application.properties:40]: same here

File: scripts/ditto_privileges.sql [hunk 15]
```diff
@@ -0,0 +1,9 @@
+USE `arrowhead`;
```
Comment (rbocsi) [inline scripts/ditto_privileges.sql:1]: you should add some lines into the create_empty_arrowhead_db.sql file to create those users that referenced here and include this file.

**Review by rbocsi — CHANGES_REQUESTED** [review]
See inline comments. After fixing those, somebody probably Svetlin can create a branch for you, and redirect this PR to the new branch.

**Comment by tsvetlin** [comment]
@nicnys-8 Base branch changed

**Pull request merged**